### ISSUE TITLE
FR: walk OIDs in parallel

### ISSUE THREAD

**Issue by Cougar** [prometheus/snmp_exporter#349]
Hi,

I have some boxes with a very big number of interfaces and collecting all the data using table OID takes minutes. I can split table to separate entries in 'walk:' list but actual walking still happens one by one in sequence and doesn't make it any faster.

Would be great if (some of) walks could be done in parallel.

I suggest to have a configuration option for number of parallel walks that defaults to 1 (current behavior).

**Comment by brian-brazil** [comment]
Is your device capable of handling multiple walks at once? Most network devices are CPU constrained and unable to do so. You should also check that the snmp exporter is close to the device network wise.

**Comment by SuperQ** [comment]
If you have a large number of walks, you could run different Prometheus jobs for different walk modules. This will be executed in parallel.

In addition to what Brian said (SNMP is very latency sensitive, run the exporter with 10ms of the devies if possible), if you're using SNMP v1, it will be much slower than v2/v3.

**Comment by RichiH** [comment]
Copying comment over from https://docs.google.com/document/d/1McJJIiJfHgoecVrVNXx4ABJmI5M21e-6O9IgMNbVnvw

> Tarko Tikan
> 2:32 PM Feb 28
> See http://p.ip.fi/g0Ss.txt (this is "old" CPU ISAM, newer generation hardware scales better but do not have box with large enough tables at hand right now). Each bulkwalk returns 720 OIDs.

**Comment by Cougar** [comment]
Do multiple modules share data too? Can we use this data for label lookups or is it necessary to walk those OIDs in every module?

In my case with hundreds of interfaces which have fixed ifIndex for ifName I saved time with very straightforward patch where I can put known values to the config file and avoid SNMP walk to these OIDs. This was much easier change than parallel collection and was helpful enough in my case right now. I described this feature in #825

**Comment by SuperQ** [comment]
No, when scraping multiple modules they currently do not share data. 

Static values are an interesting idea. Although I think a better long-term solution would be to implement some kind of SNMP cache. This has been requested before as a way to help with HA Prometheus configurations.
